README.md
# MUonESim_v2
# MUonESim_v3
# MUonESim_v19
# MUonESim_v22


History
10/12/2020 Version 1: Project created.

10/1/2021 Version 2: Energy > 1 GeV cut included for all histograms. Process ID is saved in Histogram 11, where: 	          
processID = 1; //ionization
processID = 2; //pair production
processID = 3; //Bremsstrahlung
processID = 4; //nuclear interaction

19/1/2021 Version 3: Axis range in some energy histograms increased.

26/1/2021 Version 4: 2D plot axes swapped. nTuple added.

26/1/2021 Version 5: Angles and energies calculated for all charged particles at target's exit.

27/1/2021 Version 6: Charged track multiplicity counted.

2/2/2021 Version 7: Angles and energies calculated for each process separately.

17/2/2021 Version 8: Ntuple removed.

25/3/2021 Version 14: Angle between e+ and e- from pair production added. 2D plot with energies of e+ and e-.

30/3/2021 Version 15: Test version: Trying to activate separately muNucl.

31/3/2021 Version 16: Physics list from TestEm17 added instead of the old one.

1/7/2021 Version 17: Optional changes: Graphite --> Carbon, detector thickness: 1.5 cm --> 1.3 cm

12/7/2021 Version 18: Muon energy changed from 150 GeV to 160 GeV.

26/7/2021 Version 19: Target's size changed to 10 cm X 10 cm X 1.3 cm. New physics list: FTFP_BERT_EMZ.

12/8/2021 Version 20: Debugging version: changed names of processes.

13/8/2021 Version 21: Add histograms related to muonNuclear and muBrems particles without any cuts.

15/8/2021 Assumption Version 22: Add a histograms of all processes without any cut.

16/8/2021 Version 23: More correlation plots added. Beam xy plot at the entrance and exit of the target added.

30/8/2021 Version 24: Bugs fixed from previous version.

31/8/2021 Version 25: Position of GPS changed.

1/9/2021 Version 26: In muon.mac: sigmas of GPS changed/added.

7/9/2021 Version 27: If case for filling in histogram 038 removed.

23/9/2021 Version 27a: Tests with a cut of 0.2 GeV instead of 1 GeV.


MUonESim.cc
//
// ********************************************************************
// * License and Disclaimer                                           *
// *                                                                  *
// * The  Geant4 software  is  copyright of the Copyright Holders  of *
// * the Geant4 Collaboration.  It is provided  under  the terms  and *
// * conditions of the Geant4 Software License,  included in the file *
// * LICENSE and available at  http://cern.ch/geant4/license .  These *
// * include a list of copyright holders.                             *
// *                                                                  *
// * Neither the authors of this software system, nor their employing *
// * institutes,nor the agencies providing financial support for this *
// * work  make  any representation or  warranty, express or implied, *
// * regarding  this  software system or assume any liability for its *
// * use.  Please see the license in the file  LICENSE  and URL above *
// * for the full disclaimer and the limitation of liability.         *
// *                                                                  *
// * This  code  implementation is the result of  the  scientific and *
// * technical work of the GEANT4 collaboration.                      *
// * By using,  copying,  modifying or  distributing the software (or *
// * any work based  on the software)  you  agree  to acknowledge its *
// * use  in  resulting  scientific  publications,  and indicate your *
// * acceptance of all terms of the Geant4 Software license.          *
// ********************************************************************
//
//
//
// $Id: MUonESim.cc
//
//....oooOO0OOooo........oooOO0OOooo........oooOO0OOooo........oooOO0OOooo......
//....oooOO0OOooo........oooOO0OOooo........oooOO0OOooo........oooOO0OOooo......

#include "G4RunManager.hh"
#include "G4MTRunManager.hh"
#include "G4UImanager.hh"

#include "DetectorConstruction.hh"
#include "PhysicsList.hh"
#include "PrimaryGeneratorAction.hh"
#include "RunAction.hh"
#include "EventAction.hh"
#include "SteppingAction.hh"
#include "SteppingVerbose.hh"
#include "StackingAction.hh"

#include "Randomize.hh"
#include "time.h"

#include <stdio.h>
#include <stdlib.h>

#include "G4VisExecutive.hh"
#include "G4UIExecutive.hh"


//....oooOO0OOooo........oooOO0OOooo........oooOO0OOooo........oooOO0OOooo......

int main(int argc,char** argv) {
    
    // Detect interactive mode (if no arguments) and define UI session
    //
    G4UIExecutive* ui = 0;
    if ( argc == 1 ) {
    ui = new G4UIExecutive(argc, argv);
//    ui = new G4UIExecutive(argc, argv, "tcsh");
    }


  //choose the Random engine
  CLHEP::HepRandom::setTheEngine(new CLHEP::RanecuEngine);
  //seed random generator
  G4long seed = time(NULL);
  CLHEP::HepRandom::setTheSeed(seed);


  //my Verbose output class
  G4VSteppingVerbose::SetInstance(new SteppingVerbose);
  
  // Construct the default run manager
  G4RunManager * runManager = new G4RunManager;
    
  // set mandatory initialization classes
  DetectorConstruction* detector;
  detector = new DetectorConstruction;
  runManager->SetUserInitialization(detector);
  runManager->SetUserInitialization(new PhysicsList());

  // set user action classes
  //

  //primaryGenerator
  PrimaryGeneratorAction* primary = new PrimaryGeneratorAction(detector);
  runManager->SetUserAction(primary);

  //runAction
  RunAction* runaction = new RunAction(detector, primary);
  runManager->SetUserAction(runaction);

  //eventAction
  EventAction* eventaction = new EventAction(detector, runaction);
  runManager->SetUserAction(eventaction);

  //stackAction
  StackingAction* stackingaction = new StackingAction(detector, runaction, eventaction);
  runManager->SetUserAction(stackingaction);

  //stepAction
  SteppingAction* steppingaction = new SteppingAction(detector, runaction, eventaction, stackingaction);
  runManager->SetUserAction(steppingaction);
    
    // Initialize visualization
    //
    G4VisManager* visManager = new G4VisExecutive;
    // G4VisExecutive can take a verbosity argument - see /vis/verbose guidance.
    // G4VisManager* visManager = new G4VisExecutive("Quiet");
    visManager->Initialize();

    // Get the pointer to the User Interface manager
    G4UImanager* UImanager = G4UImanager::GetUIpointer();

    // Process macro or start UI session
    //
    if ( ! ui ) {
      // batch mode
      G4String command = "/control/execute ";
      G4String fileName = argv[1];
      UImanager->ApplyCommand(command+fileName);
    }
    else {
      // interactive mode
      UImanager->ApplyCommand("/control/execute init_vis.mac");
      ui->SessionStart();
      delete ui;
    }

    // Job termination
    // Free the store: user actions, physics_list and detector_description are
    // owned and deleted by the run manager, so they should not be deleted
    // in the main() program !
    
    delete visManager;
    delete runManager;
}
//....oooOO0OOooo........oooOO0OOooo........oooOO0OOooo........oooOO0OOooo......



src/StackingAction.cc
//
// ********************************************************************
// * License and Disclaimer                                           *
// *                                                                  *
// * The  Geant4 software  is  copyright of the Copyright Holders  of *
// * the Geant4 Collaboration.  It is provided  under  the terms  and *
// * conditions of the Geant4 Software License,  included in the file *
// * LICENSE and available at  http://cern.ch/geant4/license .  These *
// * include a list of copyright holders.                             *
// *                                                                  *
// * Neither the authors of this software system, nor their employing *
// * institutes,nor the agencies providing financial support for this *
// * work  make  any representation or  warranty, express or implied, *
// * regarding  this  software system or assume any liability for its *
// * use.  Please see the license in the file  LICENSE  and URL above *
// * for the full disclaimer and the limitation of liability.         *
// *                                                                  *
// * This  code  implementation is the result of  the  scientific and *
// * technical work of the GEANT4 collaboration.                      *
// * By using,  copying,  modifying or  distributing the software (or *
// * any work based  on the software)  you  agree  to acknowledge its *
// * use  in  resulting  scientific  publications,  and indicate your *
// * acceptance of all terms of the Geant4 Software license.          *
// ********************************************************************
//
/// \file electromagnetic/TestEm18/src/StackingAction.cc
/// \brief Implementation of the StackingAction class
//
// $Id: StackingAction.cc 67268 2013-02-13 11:38:40Z ihrivnac $
//
//....oooOO0OOooo........oooOO0OOooo........oooOO0OOooo........oooOO0OOooo......
//....oooOO0OOooo........oooOO0OOooo........oooOO0OOooo........oooOO0OOooo......

#include "StackingAction.hh"

#include "DetectorConstruction.hh"
#include "EventAction.hh"
#include "HistoManager.hh"
#include "RunAction.hh"

#include "G4Step.hh"
#include "G4Track.hh"
#include "G4VTrajectory.hh"
#include "G4TransportationManager.hh"
#include "G4Navigator.hh"
#include "G4PVPlacement.hh"
#include "G4GeometryManager.hh"
#include "G4PhysicalVolumeStore.hh"
#include "G4LogicalVolumeStore.hh"
#include "G4SolidStore.hh"

#include "G4Positron.hh"
#include "G4Electron.hh"
#include "G4PionPlus.hh"
#include "G4PionMinus.hh"
#include "G4MuonPlus.hh"
#include "G4MuonMinus.hh"
#include "G4Gamma.hh"

#include "G4UnitsTable.hh"
#include "G4SystemOfUnits.hh"
#include "G4VisAttributes.hh"

#include <stdio.h>
#include <stdlib.h>
#include <time.h>

//....oooOO0OOooo........oooOO0OOooo........oooOO0OOooo........oooOO0OOooo......

StackingAction::StackingAction(DetectorConstruction* DA, RunAction* RA, EventAction* EA)
:G4UserStackingAction(), fDetectorconstruction(DA), fRunaction(RA), fEventaction(EA)
{}

//....oooOO0OOooo........oooOO0OOooo........oooOO0OOooo........oooOO0OOooo......

StackingAction::~StackingAction()
{}

//....oooOO0OOooo........oooOO0OOooo........oooOO0OOooo........oooOO0OOooo......

G4ClassificationOfNewTrack
StackingAction::ClassifyNewTrack(const G4Track* track)
{
  //define speed unit
  new G4UnitDefinition ("meter/second", "m/s", "Speed", m/s);
  
  //energy spectrum, polarization, initial momentum direction, track weight, velocity and local time of secondaries and tertiaries
  //
  energy = track->GetKineticEnergy();
  xpolarization = track->GetPolarization().x();
  ypolarization = track->GetPolarization().y();
  zpolarization = track->GetPolarization().z();
  charged = (track->GetDefinition()->GetPDGCharge() != 0.);
  G4VPhysicalVolume* volume = track->GetVolume();
  /*G4double trackWeight = track->GetWeight();
  G4double initVelocity = track->GetVelocity();
  G4double kinEnergy = track->GetKineticEnergy();
  G4double time = track->GetLocalTime();*/

  //geometry parameters
  /*G4double TargetSizeX, TargetSizeY, TargetSizeZ;
  G4double pos;
 
  TargetSizeX=fDetectorconstruction->GetSizeXT();
  TargetSizeY=fDetectorconstruction->GetSizeYT();
  TargetSizeZ=fDetectorconstruction->GetSizeZT();*/

  //G4ThreeVector primVert = track->GetVertexPosition();
  /*G4double x_prim, y_prim, z_prim;
  x_prim = track->GetVertexPosition().x();
  y_prim = track->GetVertexPosition().y();
  z_prim = track->GetVertexPosition().z();*/

  //secret number
  //G4int iSecret;

  //G4cout 
      //<< "\n Local time: " 
      //<< G4BestUnit(time, "Time")
      //<< "\n Initial velocity: " 
      //<< G4BestUnit(initVelocity, "Speed")
      //<< "\n Kinetic energy: "
      //<< G4BestUnit(kinEnergy, "Energy")
      //<< "\n Track weight: " 
      //<< trackWeight
      //<< G4endl;

  //particle definition
  const G4ParticleDefinition* particleDefinition = track->GetParticleDefinition();

  G4ThreeVector initMom = track->GetMomentum();

  //kinetic energy of secondaries calculation 
  if (track->GetParentID() == 1)  {

    if (volume == fDetectorconstruction->GetTarget()) {

      if(particleDefinition == G4Positron::Definition())   {
        fRunaction->AddNbPositronsTarget();
      }

      if(particleDefinition == G4Electron::Definition())   {
        fRunaction->AddNbElectronsTarget();
      }

      if(particleDefinition == G4MuonPlus::Definition())   {
        fRunaction->AddNbMuonsPlusTarget();
      }

      if(particleDefinition == G4MuonMinus::Definition())   {
        fRunaction->AddNbMuonsMinusTarget();
      }

    }

  }
 return fUrgent;
}

//....oooOO0OOooo........oooOO0OOooo........oooOO0OOooo........oooOO0OOooo......


src/HistoManager.cc
//
// ********************************************************************
// * License and Disclaimer                                           *
// *                                                                  *
// * The  Geant4 software  is  copyright of the Copyright Holders  of *
// * the Geant4 Collaboration.  It is provided  under  the terms  and *
// * conditions of the Geant4 Software License,  included in the file *
// * LICENSE and available at  http://cern.ch/geant4/license .  These *
// * include a list of copyright holders.                             *
// *                                                                  *
// * Neither the authors of this software system, nor their employing *
// * institutes,nor the agencies providing financial support for this *
// * work  make  any representation or  warranty, express or implied, *
// * regarding  this  software system or assume any liability for its *
// * use.  Please see the license in the file  LICENSE  and URL above *
// * for the full disclaimer and the limitation of liability.         *
// *                                                                  *
// * This  code  implementation is the result of  the  scientific and *
// * technical work of the GEANT4 collaboration.                      *
// * By using,  copying,  modifying or  distributing the software (or *
// * any work based  on the software)  you  agree  to acknowledge its *
// * use  in  resulting  scientific  publications,  and indicate your *
// * acceptance of all terms of the Geant4 Software license.          *
// ********************************************************************
//
//
// $Id: HistoManager.cc 72242 2013-07-12 08:44:19Z gcosmo $
// 
//....oooOO0OOooo........oooOO0OOooo........oooOO0OOooo........oooOO0OOooo......
//....oooOO0OOooo........oooOO0OOooo........oooOO0OOooo........oooOO0OOooo......

#include <CLHEP/Units/SystemOfUnits.h>
#include <sstream>

#include "HistoManager.hh"
#include "G4UnitsTable.hh"
#include "G4SystemOfUnits.hh"

//....oooOO0OOooo........oooOO0OOooo........oooOO0OOooo........oooOO0OOooo......

HistoManager::HistoManager()
  :fFileName("MUonESim")
{
  Book();
}

//....oooOO0OOooo........oooOO0OOooo........oooOO0OOooo........oooOO0OOooo......

HistoManager::~HistoManager()
{
  delete G4AnalysisManager::Instance();
}

//....oooOO0OOooo........oooOO0OOooo........oooOO0OOooo........oooOO0OOooo......

void HistoManager::Book()
{
  // Create or get analysis manager
  // The choice of analysis technology is done via selection of a namespace
  // in HistoManager.hh
  // Creating a tree container to handle histograms and ntuples.
  // This tree is associated to an output file.
  //
  G4AnalysisManager* analysisManager = G4AnalysisManager::Instance();
  analysisManager->SetFileName(fFileName);
  analysisManager->SetVerboseLevel(1);
  analysisManager->SetActivation(true);   //enable inactivation of histograms

  
  // Define histograms start values
  // const G4int kNbofStrips = 1016;
  const G4int kMaxHisto = 39;
  const G4int kMaxHisto2 = 16;

  const G4String id[] = { "000", "001", "002", "003", "004", "005", "006", "007", "008", "009", "010", "011", "012", "013", "014", "015", "016", "017", "018", "019", "020", "021", "022", "023", "024", "025", "026", "027", "028", "029", "030", "031", "032", "033", "034", "035", "036", "037", "038" };
  const G4String id2[] = { "0", "1", "2", "3", "4", "5", "6", "7", "8", "9", "10", "11", "12", "13", "14", "15"};

  const G4String title[] =
                { "dummy",                                                       //0
                  "Energy of primary muons at the exit of the target (simulation)",   //1
                  "Energy of secondary positrons at the exit of the target, E > 1 GeV (simulation)",   //2
                  "Energy of secondary electrons at the exit of the target, E > 1 GeV (simulation)",   //3
                  "Energy of secondary #mu^{+} at the exit of the target, E > 1 GeV (simulation)",   //4
                  "Energy of secondary #mu^{-} at the exit of the target, E > 1 GeV (simulation)",   //5
                  "Angle of primary muons at the exit of the target (simulation, smaller than 30 mrad)",   //6
                  "Angle of secondary positrons at the exit of the target, E > 1 GeV (simulation, smaller than 30 mrad)",   //7
                  "Angle of secondary electrons at the exit of the target, E > 1 GeV (simulation, smaller than 30 mrad)",   //8
                  "Angle of secondary #mu^{+} at the exit of the target, E > 1 GeV (simulation, smaller than 30 mrad)",   //9
                  "Angle of secondary #mu^{-} at the exit of the target, E > 1 GeV (simulation, smaller than 30 mrad)",   //10
                  "Process ID for the creation of secondary particles, E > 1 GeV (simulation)",  //11
                  "Energy of all charged particles at the exit of the target (simulation)",   //12
                  "Angle of all charged particles at the exit of the target (simulation)",   //13
                  "Energy of charged particles at the exit of the target, E > 0.5 GeV (simulation)",   //14
                  "Angle of charged particles at the exit of the target, E > 0.5 GeV (simulation)",   //15
                  "Energy of charged particles at the exit of the target, E > 1 GeV (simulation)",   //16
                  "Angle of charged particles at the exit of the target, E > 1 GeV (simulation)",   //17
                  "Multiplicity of charged tracks with angle < 30 mrad (simulation)",   //18
                  "Multiplicity of charged tracks with angle < 30 mrad, E > 0.5 GeV (simulation)",   //19
                  "Multiplicity of charged tracks with angle < 30 mrad, E > 1 GeV (simulation)",   //20
                  "Energy of secondary e at the exit of the target, ionization (simulation)",   //21   
                  "Energy of secondary e at the exit of the target, pair production (simulation)",   //22
                  "Energy of secondary e at the exit of the target, ionization, E > 0.5 GeV (simulation)",   //23  
                  "Energy of secondary e at the exit of the target, pair production, E > 0.5 GeV (simulation)",   //24 
                  "Energy of secondary e at the exit of the target, ionization, E > 1 GeV (simulation)",   //25 
                  "Energy of secondary e at the exit of the target, pair production, E > 1 GeV (simulation)",   //26
                  "Angle of secondary e at the exit of the target, ionization (simulation)",   //27  
                  "Angle of secondary e at the exit of the target, pair production (simulation)",   //28
                  "Angle of secondary e at the exit of the target, ionization, E > 0.5 GeV (simulation)",   //29 
                  "Angle of secondary e at the exit of the target, pair production, E > 0.5 GeV (simulation)",   //30 
                  "Angle of secondary e at the exit of the target, ionization, E > 1 GeV (simulation)",   //31 
                  "Angle of secondary e at the exit of the target, pair production, E > 1 GeV (simulation)",   //32
                  "Angle between secondary e+ and e- tracks at the exit of the target, E > 1 GeV (simulation)",   //33
                  "Energy of secondary particle at the exit of the target, Bremsstrahlung (simulation)",   //34
                  "Energy of secondary particle at the exit of the target, nuclear interaction (simulation)",   //35
                  "Angle of secondary particle at the exit of the target, Bremsstrahlung (simulation)",   //36
                  "Angle of secondary particle at the exit of the target, nuclear interaction (simulation)",    //37
                  "Process ID for the creation of secondary particles, no cuts (simulation)"  //38
                };

  const G4String title2[] =
                { "dummy",                                                     //0
                  "Angle of primary muon vs. angle of secondary positron (simulation)",   //1
                  "Angle of primary muon vs. angle of secondary electron (simulation)",   //2
                  "Angle of primary muon vs. angle of secondary #mu^{+} (simulation)",    //3
                  "Angle of primary muon vs. angle of secondary #mu^{-} (simulation)",    //4
                  "Angle of primary muon vs. angle of secondary e, ionization (simulation)",   //5
                  "Angle of primary muon vs. angle of secondary e, pair production (simulation)",   //6
                  "Angle of primary muon vs. angle of secondary e, ionization, E > 0.5 GeV (simulation)",   //7
                  "Angle of primary muon vs. angle of secondary e, pair production, E > 0.5 GeV (simulation)",   //8
                  "Angle of primary muon vs. angle of secondary e, ionization, E > 1 GeV (simulation)",   //9
                  "Angle of primary muon vs. angle of secondary e, pair production, E > 1 GeV (simulation)",   //10
                  "Energy of secondary e+ vs. energy of secondary e-, E > 1 GeV (simulation)",   //11
                  "Angle of secondary e_{1} vs. angle of secondary e_{2}, pair production, E > 1 GeV (simulation)",   //12
                  "Angle of secondary e vs. energy of secondary e, pair production, E > 1 GeV (simulation)",   //13
                  "Beam xy spot at the entrance of the target (simulation)",   //14
                  "Beam xy spot at the exit of the target (simulation)"    //15
                };
            
  // Default values (to be reset via /analysis/h1/set command)               
  G4int nbins = 100;
  G4double vmin = -0.5;
  G4double vmax = 100.5;

  // Create all histograms as inactivated 
  // as we have not yet set nbins, vmin, vmax
  for (G4int k=0; k<kMaxHisto; k++) {
    G4int ih = analysisManager->CreateH1(id[k], title[k], nbins, vmin, vmax);
    analysisManager->SetH1Activation(ih, false);
  }

  for (G4int k2=0; k2<kMaxHisto2; k2++) {
    G4int ih2 = analysisManager->CreateH2(id2[k2], title2[k2], nbins, vmin, vmax, nbins, vmin, vmax);
    analysisManager->SetH2Activation(ih2, false);
  }

  G4int ih0 = analysisManager->CreateH2("0", "dummy", nbins, vmin, vmax,  //real point
						      nbins, vmin, vmax);   //reconstructed point
  analysisManager->SetH2Activation(ih0, false);
  
  // nTuples
  /*analysisManager->SetNtupleDirectoryName("ntuple");
  analysisManager->SetFirstNtupleId(1);       
  analysisManager->CreateNtuple("1", "Tuple 1");
  
  std::ostringstream os0;
  os0 <<"Dummy";
  analysisManager->CreateNtupleDColumn(os0.str());
  
  std::ostringstream os1;
  os1 <<"EnergyPrimaryMuonExitTarget";
  analysisManager->CreateNtupleDColumn(os1.str());
  
  std::ostringstream os2;
  os2 <<"EnergySecondaryPositronExitTarget";
  analysisManager->CreateNtupleDColumn(os2.str());
  
  std::ostringstream os3;
  os3 <<"EnergySecondaryElectronExitTarget";
  analysisManager->CreateNtupleDColumn(os3.str());
  
  std::ostringstream os4;
  os4 <<"EnergySecondaryMuonPlusExitTarget";
  analysisManager->CreateNtupleDColumn(os4.str());
  
  std::ostringstream os5;
  os5 <<"EnergySecondaryMuonMinusExitTarget";
  analysisManager->CreateNtupleDColumn(os5.str());
  
  std::ostringstream os6;
  os6 <<"AnglePrimaryMuonExitTarget";
  analysisManager->CreateNtupleDColumn(os6.str());
  
  std::ostringstream os7;
  os7 <<"AngleSecondaryPositronExitTarget";
  analysisManager->CreateNtupleDColumn(os7.str());
  
  std::ostringstream os8;
  os8 <<"AngleSecondaryElectronExitTarget";
  analysisManager->CreateNtupleDColumn(os8.str());
  
  std::ostringstream os9;
  os9 <<"AngleSecondaryMuonPlusExitTarget";
  analysisManager->CreateNtupleDColumn(os9.str());
  
  std::ostringstream os10;
  os10 <<"AngleSecondaryMuonMinusExitTarget";
  analysisManager->CreateNtupleDColumn(os10.str());
  
  std::ostringstream os11;
  os11 <<"ProcessID";
  analysisManager->CreateNtupleDColumn(os11.str());
  
  std::ostringstream os12;
  os12 <<"EnergyChargedExitTarget";
  analysisManager->CreateNtupleDColumn(os12.str());
  
  std::ostringstream os13;
  os13 <<"AngleChargedExitTarget";
  analysisManager->CreateNtupleDColumn(os13.str());
  
  std::ostringstream os14;
  os14 <<"EnergyChargedCut05GeVExitTarget";
  analysisManager->CreateNtupleDColumn(os14.str());
  
  std::ostringstream os15;
  os15 <<"AngleChargedCut05GeVExitTarget";
  analysisManager->CreateNtupleDColumn(os15.str());
  
  std::ostringstream os16;
  os16 <<"EnergyChargedCut1GeVExitTarget";
  analysisManager->CreateNtupleDColumn(os16.str());
  
  std::ostringstream os17;
  os17 <<"AngleChargedCut1GeVExitTarget";
  analysisManager->CreateNtupleDColumn(os17.str());
  
  std::ostringstream os18;
  os18 <<"NbTracks";
  analysisManager->CreateNtupleDColumn(os18.str());
  
  std::ostringstream os19;
  os19 <<"NbTracksCut05GeV";
  analysisManager->CreateNtupleDColumn(os19.str());
  
  std::ostringstream os20;
  os20 <<"NbTracksCut1GeV";
  analysisManager->CreateNtupleDColumn(os20.str());
  
  std::ostringstream os21;
  os21 <<"EnergyIoni";
  analysisManager->CreateNtupleDColumn(os21.str());
  
  std::ostringstream os22;
  os22 <<"EnergyPP";
  analysisManager->CreateNtupleDColumn(os22.str());
  
  std::ostringstream os23;
  os23 <<"EnergyIoniCut05GeV";
  analysisManager->CreateNtupleDColumn(os23.str());
  
  std::ostringstream os24;
  os24 <<"EnergyPPCut05GeV";
  analysisManager->CreateNtupleDColumn(os24.str());
  
  std::ostringstream os25;
  os25 <<"EnergyIoniCut1GeV";
  analysisManager->CreateNtupleDColumn(os25.str());
  
  std::ostringstream os26;
  os26 <<"EnergyPPCut1GeV";
  analysisManager->CreateNtupleDColumn(os26.str());
  
  std::ostringstream os27;
  os27 <<"AngleIoni";
  analysisManager->CreateNtupleDColumn(os27.str());
  
  std::ostringstream os28;
  os28 <<"AnglePP";
  analysisManager->CreateNtupleDColumn(os28.str());
  
  std::ostringstream os29;
  os29 <<"AngleIoniCut05GeV";
  analysisManager->CreateNtupleDColumn(os29.str());
  
  std::ostringstream os30;
  os30 <<"AnglePPCut05GeV";
  analysisManager->CreateNtupleDColumn(os30.str());
  
  std::ostringstream os31;
  os31 <<"AngleIoniCut1GeV";
  analysisManager->CreateNtupleDColumn(os31.str());
  
  std::ostringstream os32;
  os32 <<"AnglePPCut1GeV";
  analysisManager->CreateNtupleDColumn(os32.str());
  
  analysisManager->FinishNtuple(); */
  
}

//....oooOO0OOooo........oooOO0OOooo........oooOO0OOooo........oooOO0OOooo......
